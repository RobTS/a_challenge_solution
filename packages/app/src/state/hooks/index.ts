import { useDispatch } from 'react-redux';
import { Dispatch } from 'redux';
import { useCallback } from 'react';

export const useThunk = (): (<T>(x: (dispatch: Dispatch) => T) => T) => {
  const dispatch = useDispatch();
  const wrapper = useCallback(
    <T>(thunk: (dispatch: Dispatch) => T) => {
      return thunk(dispatch);
    },
    [dispatch],
  );
  return wrapper;
};
