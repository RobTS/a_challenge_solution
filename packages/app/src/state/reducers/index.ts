import { combineReducers } from 'redux';
import { profilesReducer } from './profiles';

export const rootReducer = combineReducers({
  profiles: profilesReducer,
});
