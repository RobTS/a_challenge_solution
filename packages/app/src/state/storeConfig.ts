import { Middleware, applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import { rootReducer } from './reducers';

export const configureStore = () => {
  const middlewares: Middleware[] = [thunk];

  return createStore(rootReducer, applyMiddleware(...middlewares));
};
