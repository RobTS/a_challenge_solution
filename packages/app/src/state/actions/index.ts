import { SAFE_CONSTANTS } from '../constants';
import { action } from 'typesafe-actions';
import { Profile } from '../entities';

export const fetchDataStart = () => action(SAFE_CONSTANTS.FETCH_DATA_START);

export const fetchDataSuccess = (profiles: Profile[]) =>
  action(SAFE_CONSTANTS.FETCH_DATA_SUCCESS, profiles);

export const fetchDataFail = (error: Error) =>
  action(SAFE_CONSTANTS.FETCH_DATA_FAIL, undefined, undefined, error);
