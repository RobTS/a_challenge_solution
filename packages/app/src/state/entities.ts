import { ActionType } from 'typesafe-actions';
import * as SafeActions from './actions';

export type RequestMissingData = {
  type: 'initial' | 'loading' | 'error';
  data?: undefined;
  error?: undefined;
};

export type RequestWithData<T> = {
  type: 'success';
  data: T;
  error?: undefined;
};

export type RequestWithError = {
  type: 'error';
  error: Error;
  data?: undefined;
};

export type RequestState<T> =
  | RequestMissingData
  | RequestWithData<T>
  | RequestWithError;

export type Profile = {
  id: string;
  avatar: string;
  name: string;
};

export type SafeActionType = ActionType<typeof SafeActions>;
