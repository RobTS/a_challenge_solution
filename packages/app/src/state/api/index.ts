import { Profile } from '../entities';
import { SERVER_URL } from '../../config';

const wrappedFetch = (url: string) => {
  return fetch(SERVER_URL + url, {
    method: 'GET',
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  }).then((data) => data.json());
};

// Promise.race as timeout implementation due to lack of a native one
const wrappedFetchWithTimeout = <T>(url: string) => {
  return Promise.race([
    wrappedFetch(url) as Promise<T>,
    new Promise<Profile[]>((_, reject) =>
      setTimeout(() => reject(new Error('Timeout while fetching')), 30000),
    ),
  ]);
};

export const fetchData = (): Promise<Profile[]> => {
  return wrappedFetchWithTimeout<Profile[]>('/');
};

export const Api = {
  fetchData,
};
