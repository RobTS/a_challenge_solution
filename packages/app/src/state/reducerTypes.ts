import { Profile, RequestState } from './entities';

export type ProfileReducerState = RequestState<Profile[]>;

export type GlobalState = {
  profiles: ProfileReducerState;
};
