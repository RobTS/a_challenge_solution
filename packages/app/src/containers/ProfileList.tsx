import React, { useCallback } from 'react';
import { FlatList, StyleSheet, Text, View } from 'react-native';
import { useSelector } from 'react-redux';
import { GlobalState } from '../state/reducerTypes';
import { ProfileListItem } from '../components/ProfileListItem';

const ListHeader = () => {
  const profileRequestState = useSelector(
    (state: GlobalState) => state.profiles.type,
  );

  if (profileRequestState === 'success' || profileRequestState === 'initial') {
    return null;
  }

  return (
    <View style={styles.messageHeader}>
      {profileRequestState === 'error' ? (
        <Text style={styles.messageHeaderText}>
          No connection to the server. Have you started it and adapted the
          SERVER_URL?
        </Text>
      ) : null}
      {profileRequestState === 'loading' ? <Text>Loading content</Text> : null}
    </View>
  );
};

export const ProfileList: React.FC = () => {
  const profileRequest = useSelector((state: GlobalState) => state.profiles);

  const renderItem = useCallback(({ item, index }) => {
    return (
      <ProfileListItem
        even={index % 2 === 0}
        id={item.id}
        avatar={item.avatar}
        name={item.name}
        key={item.id}
      />
    );
  }, []);

  return (
    <FlatList
      style={styles.listView}
      contentContainerStyle={styles.listContainer}
      ListHeaderComponent={<ListHeader />}
      data={profileRequest.type === 'success' ? profileRequest.data : []}
      renderItem={renderItem}
    />
  );
};

const styles = StyleSheet.create({
  messageHeader: {
    alignItems: 'center',
    justifyContent: 'center',
    padding: 10,
    backgroundColor: '#ffa500',
  },
  messageHeaderText: {
    textAlign: 'center',
  },
  listView: { flex: 1 },
  listContainer: {
    alignItems: 'stretch',
    justifyContent: 'center',
    backgroundColor: '#ffffff',
  },
});
