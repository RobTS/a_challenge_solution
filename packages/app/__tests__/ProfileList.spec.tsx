import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { ProfileList } from '../src/containers/ProfileList';
import { configureStore } from '../src/state/storeConfig';
import {
  fetchDataFail,
  fetchDataStart,
  fetchDataSuccess,
} from '../src/state/actions';

describe('The profile list', () => {
  it('renders correctly without content', () => {
    const store = configureStore();
    const tree = renderer
      .create(
        <Provider store={store}>
          <ProfileList />
        </Provider>,
      )
      .toJSON() as any;
    expect(tree.children).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with content', () => {
    const store = configureStore();
    store.dispatch(
      fetchDataSuccess([
        {
          id: '2',
          avatar: 'https://i.pravatar.cc/150?img=2',
          name: 'Esme Coleman',
        },
      ]),
    );
    const tree = renderer
      .create(
        <Provider store={store}>
          <ProfileList />
        </Provider>,
      )
      .toJSON() as any;
    expect(tree.children).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly when loading', () => {
    const store = configureStore();
    store.dispatch(fetchDataStart());
    const tree = renderer
      .create(
        <Provider store={store}>
          <ProfileList />
        </Provider>,
      )
      .toJSON() as any;
    expect(tree.children).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly on fetch error', () => {
    const store = configureStore();
    store.dispatch(fetchDataFail(new Error('This should not happen')));
    const tree = renderer
      .create(
        <Provider store={store}>
          <ProfileList />
        </Provider>,
      )
      .toJSON() as any;
    expect(tree.children).toHaveLength(1);
    expect(tree).toMatchSnapshot();
  });
});
