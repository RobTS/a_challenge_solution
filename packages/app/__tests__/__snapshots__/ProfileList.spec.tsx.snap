// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The profile list renders correctly on fetch error 1`] = `
<RCTScrollView
  ListHeaderComponent={<ListHeader />}
  contentContainerStyle={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#ffffff",
      "justifyContent": "center",
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffa500",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          No connection to the server. Have you started it and adapted the SERVER_URL?
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`The profile list renders correctly when loading 1`] = `
<RCTScrollView
  ListHeaderComponent={<ListHeader />}
  contentContainerStyle={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#ffffff",
      "justifyContent": "center",
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#ffa500",
            "justifyContent": "center",
            "padding": 10,
          }
        }
      >
        <Text>
          Loading content
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`The profile list renders correctly with content 1`] = `
<RCTScrollView
  ListHeaderComponent={<ListHeader />}
  contentContainerStyle={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#ffffff",
      "justifyContent": "center",
    }
  }
  data={
    Array [
      Object {
        "avatar": "https://i.pravatar.cc/150?img=2",
        "id": "2",
        "name": "Esme Coleman",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignSelf": "stretch",
            "position": "relative",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "backgroundColor": "#000000",
                "flex": 1,
                "height": 421.875,
                "width": 750,
              },
            ]
          }
        >
          <FastImageView
            resizeMode="cover"
            source={
              Object {
                "priority": "normal",
                "uri": "https://i.pravatar.cc/150?img=2",
              }
            }
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
          />
        </View>
        <Text
          numberOfLines={1}
          style={
            Object {
              "backgroundColor": "#ffffff",
              "bottom": 30,
              "fontWeight": "bold",
              "left": 0,
              "paddingHorizontal": 20,
              "paddingVertical": 10,
              "position": "absolute",
            }
          }
        >
          Esme Coleman
        </Text>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`The profile list renders correctly without content 1`] = `
<RCTScrollView
  ListHeaderComponent={<ListHeader />}
  contentContainerStyle={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "#ffffff",
      "justifyContent": "center",
    }
  }
  data={Array []}
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  style={
    Object {
      "flex": 1,
    }
  }
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
    />
  </View>
</RCTScrollView>
`;
